src/tennissimulator/match/Set.java

```diff
@@ -0,0 +1,110 @@
+    private final ScoreTable scoreTable;
+    private final Game game;
+    public static final int GAMES = 3;
+    private Integer firstPlayerGamesWon;
+    private Integer secondPlayerGamesWon;
+
+    public Set(Player firstPlayer, Player secondPlayer, ScoreTable scoreTable) {
+        this.firstPlayer = firstPlayer;
+        this.secondPlayer = secondPlayer;
+        this.scoreTable = scoreTable;
+        game = new Game(firstPlayer, secondPlayer, scoreTable);
+        firstPlayerGamesWon = 0;
+        secondPlayerGamesWon = 0;
+    }
+
+    public Optional<Player> simulate() {
+        while (hasNoWinner()) {
+            updateResult(game.simulate());
+            scoreTable.updateWithGame(firstPlayerGamesWon, secondPlayerGamesWon);
+            changeServer();
+        }
+
+        return getWinner();
+    }
+
+    private void changeServer() {
+        firstPlayer.changeServer();
+        secondPlayer.changeServer();
+    }
+
+    public void resetScore() {
+        firstPlayerGamesWon = 0;
+        secondPlayerGamesWon = 0;
+    }
+
+    private boolean firstPlayerHasWonSet() {
+        return (firstPlayerGamesWon >= GAMES && firstPlayerGamesWon - secondPlayerGamesWon > 1)
+                || firstPlayerGamesWon == (GAMES + 1);
+    }
+
+    private boolean secondPlayerHasWonSet() {
+        return (secondPlayerGamesWon >= GAMES && secondPlayerGamesWon - firstPlayerGamesWon > 1)
+                || secondPlayerGamesWon == (GAMES + 1);
+    }
+
+    private void updateResult(Player player) {
+        if (player.equals(firstPlayer)) {
+            firstPlayerGamesWon++;
+        } else {
+            secondPlayerGamesWon++;
+        }
+    }
+
+    public boolean hasWinner() {
+        return firstPlayerHasWonSet() || secondPlayerHasWonSet();
+    }
+
+    public boolean hasTieBreak() {
+        return firstPlayerGamesWon == GAMES && secondPlayerGamesWon == GAMES;
+    }
+
+    private boolean hasNoWinner() {
+        return !hasWinner() && !hasTieBreak();
+    }
+
+    public Player simulateTieBreak(int pointsToWinTieBreak) {
+        scoreTable.initializeTieBreak();
+        Player serverAfterTieBreak = getServerOfFirstGameAfterTieBreak();
+        Tiebreak tieBreak = new Tiebreak(pointsToWinTieBreak, firstPlayer, secondPlayer, scoreTable);
+        Player tieBreakWinner = tieBreak.simulate();
+        updateResult(tieBreakWinner);
+        scoreTable.updateWithGame(firstPlayerGamesWon, secondPlayerGamesWon);
+        setNextServer(serverAfterTieBreak);
+
+        return tieBreakWinner;
+    }
+
+    public Optional<Player> getWinner() {
+        if (hasWinner()) {
+            return firstPlayerHasWonSet() ? Optional.of(firstPlayer) : Optional.of(secondPlayer);
+        }
+
+        return Optional.empty();
+    }
+
+    private Player getServerOfFirstGameAfterTieBreak() {
+        return firstPlayer.isServing() ? secondPlayer : firstPlayer;
+    }
+
+    private void setNextServer(Player player) {
```

Comment:
isn't it enough just to change server here without checking?

Reply:
According to tennis rules - player which served first in tiebreak should receive in first game after tiebreak.

So in this case could be situation where last server in tiebreak will be also serving after tiebreak. So server could be set incorrectly If just change it without checking.

Reply:
All right, didn't know that